File: nodes/tf_broadcaster.cpp
#include "ros/ros.h"
#include "sensor_msgs/Imu.h"
#include "tf/transform_broadcaster.h"
#include "geometry_msgs/Quaternion.h"
#include "tf/transform_datatypes.h"

void handle_imu_rotation(const sensor_msgs::Imu &msg) 
{
	// Creat tf broadcaster

	static tf::TransformBroadcaster br;
	
	tf::Transform transform;
	transform.setOrigin(tf::Vector3(0.0, 0.0, 0.0));
	
	tf::Quaternion q;
	
	geometry_msgs::Quaternion quat = msg.orientation;

	// Conversion function for quaternions
	quaternionMsgToTF(quat, q);
	tf::Matrix3x3 m(q);
	
	double roll, pitch, yaw;
	
	m.getRPY(roll, pitch, yaw);
	
	tf::Quaternion q2;
	q2.setRPY(roll, pitch, 0.0);
		
	transform.setRotation(q2.normalize());
	
	// Publish to tf
	br.sendTransform(tf::StampedTransform(transform, ros::Time::now(), "world", "base_link"));
}

int main(int argc, char **argv) 
{
	ros::init(argc, argv, "tf_broadcaster");
	ros::NodeHandle n;
	
	// Subscribe to Imu and set callback to publish tf data
	ros::Subscriber sub = n.subscribe("/imu/data", 4, &handle_imu_rotation);

	ros::spin();
	return 0;
}

File: nodes/python/tf_broadcaster.py
#!/usr/bin/env python
import roslib
roslib.load_manifest('tilt_laser')
import rospy

import tf
import sensor_msgs.msg
import geometry_msgs.msg
import math
from tf.transformations import quaternion_multiply

x = 1
y = 0
z = 0
w = 0

def quat_from_angular_vel(vel):
	global x, y, z, w
	angle = math.sqrt(vel[0]**2 + vel[1]**2 + vel[2]**2)
	if angle > 0.0:
		ret = (vel[0] * math.sin(angle/2.0)/angle,
                       vel[1] * math.sin(angle/2.0)/angle,
                       vel[2] * math.sin(angle/2.0)/angle,
                       math.cos(angle/2.0))
	else:
		ret = (0.0, 0.0, 0.0, 1.0)
	ret = quaternion_multiply((x,y,z,w), ret)
	x = ret[0]
	y = ret[1]
	z = ret[2]
	w = ret[3]
	return ret
	
def handle_imu_rotation(msg):
	br = tf.TransformBroadcaster()
	#rotation = msg.orientation
	vel = msg.angular_velocity
	hz = .01
	br.sendTransform((0,0,0), quat_from_angular_vel((vel.x*hz, vel.y*hz, vel.z*hz)), rospy.Time.now(), "laser", "world")

if __name__ == '__main__':
	rospy.init_node('tf_broadcaster')
	rospy.Subscriber('/imu/data', sensor_msgs.msg.Imu, handle_imu_rotation)
	rospy.spin()
